### PR TITLE
Horizontal and vertical image flipping

### PR DESCRIPTION
**Pull request by fkurmannucsc**
Split flip image button into flip horizontally and flip vertically, also storing metadata for whether a horizontal or vertical flip was applied.

### CODE REVIEW

File: static/js/main.js [hunk 1]
```diff
@@ -5,6 +5,8 @@ window.shark = window.shark || {};
 window.shark.info = window.shark.info || {};
 window.shark.screenshots = window.shark.screenshots || {};
 
+var screenshot_counter = 0;
```
Comment (eriq-augustine) [inline static/js/main.js]: Stash this in `window.shark.info`.

That aside, you should almost never be using `var`.
`var` will hoist declarations to the top of the function context, which is typically not what you want.
Favor `let` or `const`.

File: static/js/main.js [hunk 2]
```diff
@@ -241,17 +243,23 @@ function addScreenshot(screenshot) {
...
                 </div>
                 <div>
-                    <button onclick='flipImage("${screenshot.id}")'>Flip Image</button>
+                    <span>
+                        <button onclick='flipScreenshot("${screenshot.id}", axis="x")'>Vertical Flip</button>
+                        <button onclick='flipScreenshot("${screenshot.id}", axis="y")'>Horizontal Flip</button>
+                    </span>
                 </div>
+                <div>
+                    <button onclick='delete_screenshot("${screenshot.id}")'>Delete</button>
```
Comment (eriq-augustine) [inline static/js/main.js]: This PR also has deletion, but was not indicated anywhere.

Comment (eriq-augustine) [inline static/js/main.js]: Inconsistent snake_case for function.

File: static/js/main.js [hunk 3]
```diff
@@ -241,17 +243,23 @@ function addScreenshot(screenshot) {
...
         </div>
     `;
 
     document.querySelector('.screenshot-area').insertAdjacentHTML('beforeend', html);
 }
 
-// Flip an image in the screenshot area by adding it the the canvas flipped horizontally.
-function flipImage(screenshot_id) {
+// Flip a screenshot by adding it the the canvas flipped horizontally or vertically.
+function flipScreenshot(screenshot_id, axis='y') {
```
Comment (eriq-augustine) [inline static/js/main.js]: Missing space around `=`.
We put the space in all our code (even in Python).

File: static/js/main.js [hunk 4]
```diff
@@ -241,17 +243,23 @@ function addScreenshot(screenshot) {
                             value='${time_input_value}' />
                 </div>
                 <div>
-                    <button onclick='flipImage("${screenshot.id}")'>Flip Image</button>
+                    <span>
+                        <button onclick='flipScreenshot("${screenshot.id}", axis="x")'>Vertical Flip</button>
```
Comment (eriq-augustine) [inline static/js/main.js]: Does this line actually work?
Sending an argument by name is not universally supported:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters

I would just not bother making it optional.

File: static/js/main.js [hunk 5]
```diff
@@ -260,14 +268,31 @@ function flipImage(screenshot_id) {
     canvas.height = img.naturalHeight;
 
     let context = canvas.getContext('2d');
-    context.scale(-1, 1);
-    context.drawImage(img, -img.naturalWidth, 0);
 
-    // Draw the canvas to the image area.
-    img.src = canvas.toDataURL('image/jpeg');
+    // Horizontal flip.
+    if (axis === 'y') {
```
Comment (eriq-augustine) [inline static/js/main.js]: I see a lot of duplicate code here.
You should be able to clean it up so that you just tweak some values for each case and use most of the same code.

File: static/js/main.js [hunk 6]
```diff
@@ -338,7 +363,10 @@ function takeVideoScreenshot(query, xPercent, yPercent, widthPercent, heightPerc
 
     let id = randomHex();
 
-    let index_string = String(Object.keys(window.shark.screenshots).length).padStart(3, '0');
+    // Get screenshot counter value and update counter.
+    let index_string = String(screenshot_counter).padStart(3, '0');
+    screenshot_counter += 1;
```
Comment (eriq-augustine) [inline static/js/main.js]: js has `++`.
Always prefer `++` when available.

File: static/js/main.js [hunk 7]
```diff
@@ -375,6 +404,15 @@ function takeScreenshot(source, x, y, width, height, format = 'image/jpeg') {
     return canvas.toDataURL(format);
 }
 
+// Delete a screenshot, it is no longer displayed or saved.
```
Comment (eriq-augustine) [inline static/js/main.js]: I'm not a fan of this comment because it is in a bit of an in-between space.
The purpose if pretty straightforward, so you could get away without a comment.
But since you have one and enumerate how the screenshot will be treated (not displayed or saved),
then you raise questions about what else is done with it.

Users may think: "Oh, there's a comment on this function and not the other functions, so there is probably a special situation. ... Humm... if it is not displayed or saved, what happens with the screenshot? It is still in the data but just not saved to the server?"

**Review by eriq-augustine — CHANGES_REQUESTED** [review]
Good start.

File: static/js/main.js (hunk 4 above)
Comment (fkurmannucsc) [reply static/js/main.js]: Ok, changed it, it was working on Chrome.

Comment (eriq-augustine) [reply static/js/main.js]: Remember that all comments should be resolved or commented on if it needs further discussion.

File: static/js/main.js (hunk 2 above)
Comment (fkurmannucsc) [reply static/js/main.js]: Yeah, sorry, mentioned it in Slack, didn't realize I needed to do a different branch for a different PR until after committing.

File: static/js/main.js [hunk 8]
```diff
@@ -4,6 +4,7 @@ window.shark = window.shark || {};
 
 window.shark.info = window.shark.info || {};
 window.shark.screenshots = window.shark.screenshots || {};
+window.shark.screenshot_counter = 0;
```
Comment (eriq-augustine) [inline static/js/main.js]: This is a bit off.
I said store it in `window.shark.info`.

File: static/js/main.js [hunk 9]
```diff
@@ -241,17 +242,23 @@ function addScreenshot(screenshot) {
...
         </div>
     `;
 
     document.querySelector('.screenshot-area').insertAdjacentHTML('beforeend', html);
 }
 
-// Flip an image in the screenshot area by adding it the the canvas flipped horizontally.
-function flipImage(screenshot_id) {
+// Flip a screenshot by adding it the the canvas flipped horizontally or vertically.
+function flipScreenshot(screenshot_id, axis) {
```
Comment (eriq-augustine) [inline static/js/main.js]: Let's change axis to a boolean like `flip_vertical`.
Since there are only two possible options for flipping, I think a boolean would be more clean (and eliminate error cases).

File: static/js/main.js [hunk 10]
```diff
@@ -260,14 +267,25 @@ function flipImage(screenshot_id) {
...
+        direction = 'flipped_horizontally';
+        context.scale(-1, 1);
+        context.drawImage(img, -img.naturalWidth, 0);
+    } else if (axis === 'x') {
+        direction = 'flipped_vertically';
+        context.scale(1, -1);
+        context.drawImage(img, 0, -img.naturalHeight);
+    } else {
+        throw("Invalid flip axis.")
+    }
```
Comment (eriq-augustine) [inline static/js/main.js]: I still think we can do better with sharing code.
You did do good pulling out some code, but the problem is that non-trivial things are sill not shared.
Both `context.scale()` and `context.drawImage()` can be tricky, so we would prefer if they were only done in one place.

What about setting variables like `scaleX`, `scaleY`, `drawX`, `drawY` according to the direction.
(You can even be a cool kid and merge it with the above comment about booleans by defaulting the variables to one value and then have a single if branch that changes for the other flip direction.)

File: static/js/main.js [hunk 11]
```diff
@@ -338,7 +356,10 @@ function takeVideoScreenshot(query, xPercent, yPercent, widthPercent, heightPerc
 
     let id = randomHex();
 
-    let index_string = String(Object.keys(window.shark.screenshots).length).padStart(3, '0');
+    // Get screenshot counter value and update counter.
+    let index_string = String(window.shark.screenshot_counter).padStart(3, '0');
+    window.shark.screenshot_counter ++;
```
Comment (eriq-augustine) [inline static/js/main.js]: `++` should not have a space before it.
(This is generally true with all unary operators, e.g., `!`.)

(This is a pretty wild style. Have you been taught or seen it before?)

File: static/js/main.js [hunk 12]
```diff
@@ -375,6 +397,14 @@ function takeScreenshot(source, x, y, width, height, format = 'image/jpeg') {
     return canvas.toDataURL(format);
 }
 
+function deleteScreenshot(screenshot_id) {
+    // Find screenshot area to remove.
+    let screenshot_area = document.querySelector(`.screenshot[data-id="${screenshot_id}"]`);
+    screenshot_area.remove();
```
Comment (eriq-augustine) [inline static/js/main.js]: I would merge this to a one-liner.
Since `area` is not used elsewhere and we already established the convention of operating directly off a `querySelector()`, it should be fine.

**Review by eriq-augustine — CHANGES_REQUESTED** [review]
Getting closer.

File: static/js/main.js (hunk 8 above)
Comment (fkurmannucsc) [reply static/js/main.js]: Sorry about not following that direction closely.

File: static/js/main.js (hunk 11 above)
Comment (fkurmannucsc) [reply static/js/main.js]: Never been told about ++ spacing formally but it makes sense given it's a unary operator and I'd never put a space between other unary operators like !.

File: static/js/main.js [hunk 13]
```diff
@@ -241,33 +242,49 @@ function addScreenshot(screenshot) {
...
+    let direction = 'flipped_horizontally';
+    let scale_x = -1;
+    let scale_y = 1;
+    let draw_x = -img.naturalWidth;
+    let draw_y = 0;
+
+    if (flip_vertical) {
+        direction = 'flipped_vertically';
+        scale_x = 1, scale_y = -1;
+        draw_x = 0, draw_y = -img.naturalHeight;
```
Comment (eriq-augustine) [inline static/js/main.js]: In our style (most styles I run into), we don't want to set multiple variables on the same line.
(Sometimes this is relaxed for declarations (e.g. `let a = b = 1;`), but not for already defined variables. I never relax this for declarations as I think it makes it easy to miss things (and some languages like C++ can have tricky comma operators)).

File: static/js/main.js [hunk 14]
```diff
@@ -375,6 +396,13 @@ function takeScreenshot(source, x, y, width, height, format = 'image/jpeg') {
     return canvas.toDataURL(format);
 }
 
+function deleteScreenshot(screenshot_id) {
+    // Find screenshot area to remove.
```
Comment (eriq-augustine) [inline static/js/main.js]: I would remove this comment now, since the line both finds and removes.

File: static/js/main.js [hunk 15]
```diff
@@ -375,6 +396,13 @@ function takeScreenshot(source, x, y, width, height, format = 'image/jpeg') {
...
+    document.querySelector(`.screenshot[data-id="${screenshot_id}"]`).remove();
+
```
Comment (eriq-augustine) [inline static/js/main.js]: I would probably remove this whitespace and let this visual block conceptually be be "removing the screenshot".
You can keep it as two visual blocks ("removing from the dom" and "removing from the data").
Either is fine.

Note that I am only making this comment since there are other comments on this pass.
This comment would not be something that is worthy of another PR pass by itself.

**Review by eriq-augustine — CHANGES_REQUESTED** [review]
Shared code is looking real good.
Just some styling.

**Review by eriq-augustine — APPROVED** [review]
lgtm
